### PR TITLE
Support update site promotion via a declarative Jenkins pipeline

### PR DESCRIPTION
**Pull request by merks**
Improve the bundle names as they appear in the update sites. Upgrade to latest Tycho release.
Upgrade to the latest jarsigner.
Provide a functional Oomph setup, including a setup configuration and use it in
the CONTRIBUTING.md.
Ensure that the nightly target works.
Reduce workspace warnings.

https://github.com/eclipse/Xpect/issues/328
https://github.com/eclipse/Xpect/issues/330

### CODE REVIEW

**Comment by merks** [comment]
@cdietrich  @tjeske 

Guys, I see that for my safety and comfort the PR build doesn't actually use my modified Jenkinsfile.  

I've switched to use Tycho 4.0.3 which required Maven 3.9.x and Java 17, so this just don't work in the PR build.

But I can assure you that the builds really do work and demonstrated here:

https://ci.eclipse.org/oomph/job/internal/job/xpect/job/master/

How best can be get past this impasse?

If someone merges these changes,  I'll be able to modify the Oomph project catalog to point at the new setup and I can help fix any problems that may arise.  I have some time today and some time tomorrow morning.  Then my time gets more spotty...

Note that these changes are tested that the nightly target platform works again...

**Comment by cdietrich** [comment]
pushed to https://github.com/eclipse/Xpect/tree/merks-issue-328 so that jenkins actually builds it

File: Jenkinsfile [hunk 1]
```diff
@@ -1,117 +1,151 @@
...
+                '''
+            }
+          }
         }
+      }
     }
-}
+  }
 
+  post {
```
Comment (cdietrich) [inline Jenkinsfile:123]: i assume we want to keep matrix here

File: Jenkinsfile [hunk 2]
```diff
@@ -1,117 +1,151 @@
...
-            wrap([$class: 'Xvnc', useXauthority: true]) {
-                stage('test with Eclipse 2023-09 and Xtext nighly') {
-                    try{
-                        sh "JAVA_HOME=${java17Home} ${mvnHome}/bin/mvn -P!xtext-examples -Dtycho-version=2.7.5 -Dtarget-platform=eclipse_2023_09-xtext_nightly ${mvnParams} clean integration-test"
-                    }finally {
-                        junit '**/TEST-*.xml'
-                    }
-                }
-            }
+  triggers {
```
Comment (cdietrich) [inline Jenkinsfile:6]: can we make a build with both variants run daily?

File: Jenkinsfile [hunk 3]
```diff
@@ -1,117 +1,151 @@
...
+      defaultValue: true,
+      description: '''
+        Choose whether or not the bundles will be signed.
+        This is relevant only for nightly and milestone builds.
+      '''
+    )
+
+    booleanParam(
+      name: 'PROMOTE',
+      defaultValue: true,
```
Comment (cdietrich) [inline Jenkinsfile:56]: do we want this on all branches?

File: Jenkinsfile (hunk 1 above)
Comment (merks) [reply Jenkinsfile:123]: I don't know what that means.

File: Jenkinsfile (hunk 2 above)
Comment (merks) [reply Jenkinsfile:6]: Yes, but let's do that as a next step, which I'll do immediately after the merge.

File: Jenkinsfile (hunk 3 above)
Comment (merks) [reply Jenkinsfile:56]: If you look at the later logic, if only allows the master branch to promote.

**Comment by cdietrich** [comment]
https://ci.eclipse.org/xpect/job/Xpect/job/merks-issue-328/1/console

File: Jenkinsfile (hunk 3 above)
Comment (merks) [reply Jenkinsfile:56]: This part ensure that only master can produce signed content (so no user can create a PR with changes that are signed) and only when signing is enabled, only on the master branch is the promotion parameter respected.
```
          if (env.BRANCH_NAME == 'master') {
            // Only sign the master branch.
            env.ECLIPSE_SIGN = params.ECLIPSE_SIGN
          } else {
            env.ECLIPSE_SIGN =  false
          }

            // Only promote signed builds.
          env.PROMOTE = params.PROMOTE && (env.ECLIPSE_SIGN == 'true')
```

Comment (cdietrich) [reply Jenkinsfile:56]: i see.

Comment (cdietrich) [reply Jenkinsfile:56]: i wonder if we need to add a target platform = older one as well.
as xtext in contrast to emf cannot generate against old runtimes
(no issue as of now as no changes between Xtext versions currently)

File: Jenkinsfile (hunk 1 above)
Comment (merks) [reply Jenkinsfile:123]: Ah, all that postAlways() stuff that I don't understand.  I assume we want that in, but that's just copy and paste, right?  And then called from the "always{}" block.    I don't think I could test this on my own instance...

File: Jenkinsfile (hunk 3 above)
Comment (merks) [reply Jenkinsfile:56]: These are the only two defined in the repository:
```
    choice(
      name: 'TARGET_PLATFORM',
      choices: [ 'eclipse_2023_03-xtext_2_31_0', 'eclipse_2023_09-xtext_nightly' ],
```
Maybe we don't want a choice but rather build them both (or all if there is more than one).  Though I wonder what will happen if we start using new bundle names...

Comment (merks) [reply Jenkinsfile:56]: Note that for EMF I just trigger my build with different target platform choices:

https://ci.eclipse.org/emf/job/all-target-platforms/

But I only do that occasionally...

Certainly there are different design options here....

File: Jenkinsfile (hunk 1 above)
Comment (cdietrich) [reply Jenkinsfile:123]: exactly.
we use eclipse chat for build notificiations
see
https://chat.eclipse.org/#/room/#modeling-xpect-build:matrix.eclipse.org

Comment (merks) [reply Jenkinsfile:123]: That looks cool.  Shall I try to amend my commit and force push it to add this?

File: Jenkinsfile (hunk 3 above)
Comment (cdietrich) [reply Jenkinsfile:56]: the question is how actively people keep an eye.
> Maybe we don't want a choice but rather build them both (or all if there is more than one). Though I wonder what will happen if we start using new bundle names...

i assume for these we might need two options deps or so.
but i did not look yet how actively xpect uses classgraph

Comment (cdietrich) [reply Jenkinsfile:56]: and yes right now we run both.
but deploy only the result of the first target platform

**Comment by cdietrich** [comment]
both builds with the targets worked.
the promotion we will see only when merged.

are you going to push something for matrix?

**Comment by merks** [comment]
I assume you have to work your magic again for this force push...

**Comment by cdietrich** [comment]
yes. pushed again

**Review by cdietrich — APPROVED** [review]

**Comment by merks** [comment]
I force pushed again because I saw a warning about the test reports.  So I changed that to use the pattern of the older script...

**Comment by merks** [comment]
Hopefully this is the last change!

**Comment by cdietrich** [comment]
pushed your branch again

**Comment by cdietrich** [comment]
test results published too. ready to merge?

**Comment by merks** [comment]
The tests results warning is gone.  I can't tell if the matrix thing is kicking in. 

It should:

image

**Comment by merks** [comment]
READY!

Then I can change the Oomph project catalog and do a bit more testing!  Woo hoo!

**Comment by cdietrich** [comment]
good question.
in xtext we dont have an extra function.
i introduced this cause it was to only way to get it running with old pipeline.
at the same time it also just reports if build result changes.

=> can you move it to cleanup as in Xtext and mwe
https://github.com/eclipse/xtext/blob/4dd77f11cf279f2ede8036adb5f951315b1e9b35/Jenkinsfile#L99C7-L99C7
https://github.com/eclipse/mwe/blob/62067fa5f0a40dc58cfb4cf38a0f9e0ce98809e8/Jenkinsfile#L225

**Comment by cdietrich** [comment]
always wont be called on abort as it seems.
started again and the success is shown
grafik

**Comment by merks** [comment]
I see.  I will make that change.  Also I'll add some echo so that we can see what's checking and doing.

**Comment by cdietrich** [comment]
works as expected. maybe thanks. gonna push the button

**Comment by cdietrich** [comment]
many thanks @merks

**Comment by merks** [comment]
Likewise many thanks to you too!!!

**Pull request merged**